frontend/src/OneBuilding/index.jsx

```diff
@@ -0,0 +1,189 @@
+import AcademyDetail from './AcademyDetail';
+import FarmDetail from './FarmDetail';
+import MineDetail from './MineDetail';
+
+const styles = StyleSheet.create({
+  background: {
+    flex: 1,
+    backgroundColor: 'rgba(0,0,0,0.6)',
+    justifyContent: 'center',
+    alignItems: 'center',
+  },
+  container: {
+    width: 300,
+    backgroundColor: 'rgb(202,202,202)',
+    borderRadius: 6,
+  },
+  mainBody: {
+    padding: 15,
+    paddingTop: 0,
+  },
+  loading: {
+    flex: 1,
+    justifyContent: 'center',
+    alignItems: 'center',
+  },
+});
+
+function getDetailInfo(
+  type,
+  troops,
+  gold,
+  food,
+  goldGenerateRate,
+  foodGenerateRate,
+  buildingLevel,
+  createTroopGoldCost,
+  createTroopTimeCost,
+  upgradeAcademyGoldCost,
+  upgradeAcademyTimeCost,
+) {
+  switch (type) {
+    case 'Townhall':
+      return <TownhallDetail troops={troops} gold={gold} food={food} />;
+    case 'Academy':
+      return (
+        <AcademyDetail
+          buildingLevel={buildingLevel}
+          createTroopGoldCost={createTroopGoldCost}
+          createTroopTimeCost={createTroopTimeCost}
+          upgradeAcademyGoldCost={upgradeAcademyGoldCost}
+          upgradeAcademyTimeCost={upgradeAcademyTimeCost}
+        />
+      );
+    case 'Farm':
+      return <FarmDetail foodGenerateRate={foodGenerateRate} />;
+    case 'Mine':
+      return <MineDetail goldGenerateRate={goldGenerateRate} />;
+    default:
+      return null;
+  }
+}
+
+function OneBuilding({
+  isVisible, onClick, targetBuildingId, getIconImage,
+}) {
+  const oneBuildingInfo = useSelector((state) => state.oneBuilding.oneBuildingInfo);
+  const isLoading = useSelector((state) => state.oneBuilding.isLoading);
+  const error = useSelector((state) => state.oneBuilding.error);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (targetBuildingId !== -1) {
+      dispatch(fetchOneBuilding(targetBuildingId));
+    }
+  }, [targetBuildingId]);
+
+  const listOfTroops = useSelector((state) => state.troops.listOfTroops);
+  useEffect(() => {
+    dispatch(fetchTroops());
+  }, []);
+  const totalNumOfTroops = listOfTroops.length;
+
+  const mockedGold = 60;
+  const mockedFood = 60;
+  const mockedGoldGenerateRate = 20;
+  const mockedFoodGenerateRate = 30;
+
+  const mockedRules = {
+    createTroopGoldCost: 10,
+    createTroopTimeCost: 60,
+    upgradeAcademyGoldCost: 200,
+    upgradeAcademyTimeCost: 150,
+  };
+
+  if (error) {
+    return (
+      <Text>{`Oops, ${error.message}`}</Text>
+    );
+  }
+  if (isLoading) {
+    return (
+      <Modal
+        animationType="fade"
+        transparent
+        visible={isVisible}
+        presentationStyle="overFullScreen"
+        onRequestClose={onClick}
+      >
+        <View style={styles.loading}>
+          <ActivityIndicator size="large" color={colors.tealColor} />
+        </View>
+      </Modal>
+    );
+  }
+  if (Object.keys(oneBuildingInfo).length === 0) {
+    return null;
+  }
+
+  return (
+    <Modal
```

Comment:
please create a separate modal component and put it into the `common/components` folder so it can be reusable
it should render components dinamically based on what do you pass to it as its children
please check the `CardView` for example
in this case the getDetailInfo is unnecessary, since you could move this switch-case logic into to modal

something like this:

```suggestion
  let buildingComponent = null;
  if (buildingtype === 'Farm') {
    buildingComponent = <Farm param1 param2 param3 />
  } else if (buildingtype === 'Mine') {
    buildingComponent = <Mine param1 param2 param3 />
  } 
  // ...

  return (
    <Modal>
      {buildingComponent}
    </Modal>
  );
```

the main problem here is the complexity of the `getDetailInfo` function, mainly with the amount of params it gets: 11, although based on Clean Code as we discussed multiple times, it should be maximum 1-2, maybe very few, but 11 is just way too much, it will be almost impossible to test this at this moment

Reply:
also if you make it reusable, we can show the error message in a popup too in the future, or even if something goes wrong here too